src/navigation/search.py

```diff
@@ -1,46 +1,98 @@
+from __future__ import annotations
 from typing import List
+from black import color_diff
 
 import numpy as np
 
 from context import Context
 from state import BaseState
+from dataclasses import dataclass
+from drive import get_drive_command
 
+STOP_THRESH = 0.2
+DRIVE_FWD_THRESH = 0.95
 
-def gen_square_spiral_search_pattern(center: np.ndarray, spacing: float, coverage_radius: float) -> List[np.ndarray]:
-    """
-    Generates a square spiral search pattern around a center position, assumes rover is at the center position
-    :param center:          position to center spiral on (np.ndarray),
-    :param spacing:         distance in between each spiral (float)
-    :param coverage_radius: spiral covers at least this far out
-    :return:                list of positions for the rover to traverse List(np.ndarray)
-    """
-    # TODO: refactor
-    out = []
-    cur = center
-    cur_segment_len = spacing / 2
-    directions = [
-        np.array([1, 0]),
-        np.array([0, -1]),
-        np.array([-1, 0]),
-        np.array([1, 0]),
-    ]
-    direction_index = 0
-    while cur_segment_len <= coverage_radius:
-        next_point = cur + (cur_segment_len * directions[direction_index])
-        out.append(next_point)
-        cur_segment_len += spacing / 2
-        direction_index = (direction_index + 1) % len(directions)
-    return out
+
+@dataclass
+class SearchTrajectory:
+    # coordinates
+    coordinates: np.ndarray
+    # Associated fiducial for this trajectory
+    fid_id: int
+    # pt
+    cur_pt: int = 0
+
+    def get_cur_pt(self) -> np.ndarray:
+        return self.coordinates[self.cur_pt]
+
+    def inc_pt(self) -> bool:
+        """
+        Increments the tracked point in the trajectory, returns true if
+        the trajectory is finished
+        """
+        self.cur_pt += 1
+        return self.cur_pt >= len(self.coordinates)
+
+    @classmethod
+    def spiral_traj(cls, center: np.ndarray, num_turns: int, distance: int, fid_id: int) -> SearchTrajectory:
+        """
+        Generates a square spiral search pattern around a center position, assumes rover is at the center position
+        :param center:      position to center spiral on (np.ndarray)
+        :param num_turns:   number of times to spiral around
+        :param distance:    initial distance and increment (int)
+        :return:            list of positions for the rover to traverse List(np.ndarray)
+        """
+        # First we will attempt to create the "delta" vectors that get added add each point
+        # in the spiral to get to the next.
+        dirs = np.array([[0, -1], [-1, 0], [0, 1], [1, 0]])
```

Comment:
Should we make `dirs` a variable of class `SearchTrajectory` rather than created in the function?